include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,544 @@
+        }
+    }
+
+    /* Generating functions */
+    result_type
+    operator()()
+    {
+        result_type __ret = generate_internal<internal::type_traits_t<result_type>::num_elems>();
+        return __ret;
+    }
+    /* operator () overload for result portion generation */
+    result_type
+    operator()(unsigned int __random_nums)
+    {
+        result_type __ret = generate_internal<internal::type_traits_t<result_type>::num_elems>(__random_nums);
+        return __ret;
+    }
+
+    /* Shift the counter only forward relative to its current position */
+    void
+    discard(unsigned long long __z)
+    {
+        discard_internal(__z);
+    }
+
+    /* Equality operators */
+    friend bool
+    operator==(const philox_engine& __x, const philox_engine& __y)
+    {
+        return (std::equal(__x.state_.X.begin(), __x.state_.X.end(), __y.state_.X.begin()) &&
+                std::equal(__x.state_.K.begin(), __x.state_.K.end(), __y.state_.K.begin()) &&
+                std::equal(__x.state_.Y.begin(), __x.state_.Y.end(), __y.state_.Y.begin()) &&
+                __x.state_.idx == __y.state_.idx);
+    }
+    friend bool
+    operator!=(const philox_engine& __x, const philox_engine& __y)
+    {
+        return !(__x == __y);
+    }
+
+    /* Inserters and extractors */
+    template <typename __CharT, typename __Traits, typename __UIntType, std::size_t __w, std::size_t __n, std::size_t __r,
+              internal::element_type_t<__UIntType>... __consts>
+    friend std::basic_ostream<__CharT, __Traits>&
+    operator<<(std::basic_ostream<__CharT, __Traits>&, const philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+    template <typename __UIntType, std::size_t __w, std::size_t __n, std::size_t __r, internal::element_type_t<__UIntType>... __consts>
+    friend const sycl::stream&
+    operator<<(const sycl::stream&, const philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+    template <typename __CharT, typename __Traits, typename __UIntType, std::size_t __w, std::size_t __n, std::size_t __r,
+              internal::element_type_t<__UIntType>... __consts>
+    friend std::basic_istream<__CharT, __Traits>&
+    operator>>(std::basic_istream<__CharT, __Traits>&, philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+  private:
+    /* Internal generator state */
+    struct state
+    {
+        std::array<scalar_type, word_count> X;     // counters
+        std::array<scalar_type, word_count / 2> K; // keys
+        std::array<scalar_type, word_count> Y;     // results
+        scalar_type idx;                           // index
+    } state_;
+
+    /* __word_mask<_W> - scalar_type with the low _W bits set */
+    template <std::size_t _W>
+    static constexpr scalar_type __word_mask = _W ? (~scalar_type(0) >> (std::numeric_limits<scalar_type>::digits - _W)) : 0;
+
+    /* Processing mask */
+    static constexpr auto in_mask = __word_mask<word_size>;
```

Comment:
this can be merged with the previous one if _w is not used

Reply:
It's used, please see the comment above.